homeassistant/components/light/mysensors.py

```diff
@@ -109,118 +117,268 @@ def device_state_attributes(self):
             self.gateway.set_child_value(
                 self.node_id, self.child_id, set_req.V_LIGHT, 1)
 
-        if ATTR_BRIGHTNESS in kwargs and set_req.V_DIMMER in self._values and \
+        if self.gateway.optimistic:
+            # optimistically assume that light has changed state
+            self._state = True
+
+    def _turn_on_dimmer(self, **kwargs):
+        """Turn on dimmer child device."""
+        set_req = self.gateway.const.SetReq
+        brightness = self._brightness
+
+        if ATTR_BRIGHTNESS in kwargs and \
                 kwargs[ATTR_BRIGHTNESS] != self._brightness:
             brightness = kwargs[ATTR_BRIGHTNESS]
             percent = round(100 * brightness / 255)
             self.gateway.set_child_value(
                 self.node_id, self.child_id, set_req.V_DIMMER, percent)
 
-        if float(self.gateway.version) >= 1.5:
+        if self.gateway.optimistic:
+            # optimistically assume that light has changed state
+            self._brightness = brightness
+
+    def _turn_on_rgb_and_w(self, hex_template, **kwargs):
+        """Turn on RGB or RGBW child device."""
+        rgb = self._rgb
+        white = self._white
 
-            if ATTR_RGB_WHITE in kwargs and \
-                    self.value_type in (set_req.V_RGB, set_req.V_RGBW) and \
-                    kwargs[ATTR_RGB_WHITE] != self._white:
-                white = kwargs[ATTR_RGB_WHITE]
-
-            if ATTR_RGB_COLOR in kwargs and \
-                    self.value_type in (set_req.V_RGB, set_req.V_RGBW) and \
-                    kwargs[ATTR_RGB_COLOR] != self._rgb:
-                rgb = kwargs[ATTR_RGB_COLOR]
-                if set_req.V_RGBW == self.value_type:
-                    hex_template = '%02x%02x%02x%02x'
-                    color_list = rgb.append(white)
-                if set_req.V_RGB == self.value_type:
-                    hex_template = '%02x%02x%02x'
-                    color_list = rgb
-                hex_color = hex_template % tuple(color_list)
-                self.gateway.set_child_value(
-                    self.node_id, self.child_id, self.value_type, hex_color)
+        if ATTR_RGB_WHITE in kwargs and \
+                kwargs[ATTR_RGB_WHITE] != self._white:
+            white = kwargs[ATTR_RGB_WHITE]
+
+        if ATTR_RGB_COLOR in kwargs and \
+                kwargs[ATTR_RGB_COLOR] != self._rgb:
+            rgb = kwargs[ATTR_RGB_COLOR]
+            if white is not None and hex_template == '%02x%02x%02x%02x':
+                rgb.append(white)
+            hex_color = hex_template % tuple(rgb)
+            self.gateway.set_child_value(
+                self.node_id, self.child_id, self.value_type, hex_color)
 
         if self.gateway.optimistic:
             # optimistically assume that light has changed state
-            self._state = True
             self._rgb = rgb
-            self._brightness = brightness
             self._white = white
+
+    def _turn_on(self):
```

Comment:
This method needs either better naming or should just be inlined in the other methods.